Inline File Renaming

michaelkrar opened the pull request:
Closes #150 
Todo: Fix issue where clicking rename directory button while that directory is not currently being renamed toggles the directory open/close.

curran commented:
This may be of interest as a reference https://github.com/vizhub-core/vizhub/blob/01cadfb78a2611df32f981b1fd8136b70447de9e/vizhub-v2/packages/neoFrontend/src/pages/VizPage/Body/Editor/FilesSection/index.js#L14

curran commented on src/client/Sidebar/Item.jsx:
```diff
@@ -1,35 +1,76 @@
+      event.target.blur();
+    }
+  };
+  const onBlur = (event) => {
```
```suggestion
  const onBlur = useCallback((event) => {
```

curran commented on src/client/Sidebar/Item.jsx:
```diff
@@ -1,35 +1,76 @@
       className="full-box file-or-directory"
-      onClick={handleClick}
+      onClick={
+        isRenaming ? console.log(isRenaming) : handleClick
```
This line doesn't look right. Perhaps WIP to be cleaned up?

curran commented on src/client/Sidebar/Item.jsx:
```diff
@@ -1,35 +1,76 @@
+            value={renameValue}
+            onKeyDown={onKeyDown}
+            onBlur={onBlur}
+            onChange={(event) =>
```
Best to move this to a separate definition outside the JSX and use `useCallback`.

curran commented on src/client/Sidebar/Item.jsx:
```diff
@@ -1,35 +1,76 @@
             className="bx bxs-edit utilities"
             style={{ color: '#abdafb' }}
-            onClick={handleRenameClick}
+            onClick={() => {
```
Best to move this to a separate definition outside the JSX and use `useCallback`.

curran commented on src/client/Sidebar/index.tsx:
```diff
@@ -63,7 +66,7 @@ export const Sidebar = ({
               className="bx bxs-file-plus new-btn"
               style={{ color: '#dbdde1' }}
-              onClick={createFile}
+              onClick={() => createFile('New File')}
```
Best to move this to a separate definition outside the JSX and use `useCallback`. Or just change `createFile` to default the name.

curran reviewed (changes requested):
Excellent work! The `useCallback` usage would be great to add in here. Here's a nice read on the "why" behind `useCallback` https://medium.com/@KTAsim/react-performance-event-handlers-using-usecallback-hook-9e4a06f8bb2f

Alternatively, if you are able to get it to work nicely, let me know and I can take over for a final cleanup pass including addition of `useCallback`. Thanks so much for your work on this!

curran commented:
This change may help here https://github.com/vizhub-core/vzcode/pull/182/files#diff-017595a8c9d711c40f5137aa936038881760cacffebe8c98ded565b9302dee99R34 - there's a possibility that the rename function was not working properly before because of this missing dependency. Maybe pull this change into this branch once it lands, hopefully that will help get this inline file renaming feature working properly.

curran commented:
Let me know if you want to tap out @michaelkrar , I can potentially take over this PR to land it if you're out of energy for it (I know what that's like - sometimes it's tough to get it over the finish line). The bulk of the feature is done and you've done amazing work here. Thanks!

curran commented:
This PR is getting stale, some merge conflicts are arising.

Also I was thinking, how can we get this merged? We need to maintain working order. One idea: disable the rename UI entirely for directories, then re-enable it at a future time to close #103 .

michaelkrar commented:
Hi sorry for the delay I have not been checking my email for github notifications. Going on I will make sure to respond faster. I temporarily undid the default file name feature and cleaned up the changes. I also changed the two lambda functions I wrote in Item.jsx to be useCallback. Let me know if anything else needs to be changed before merge.

curran commented:
Amazing! Thank you so much for your work on this. I'll take it from here re: final cleanup and merge.

curran commented:
Notes on local testing:

 * There's a strange behavior of when you are typing something, then if you move the mouse outside the text area, it switches back to a regular file listing
 *

curran commented on src/client/Sidebar/Item.tsx:
```diff
@@ -7,13 +7,57 @@ import {
+  });
+  const onBlur = useCallback((event) => {
+    if (event.target.value.trim() === '') {
+      setEditingValue(value);
```
This doesn't exist

curran commented:
Hooray! Inline File Renaming landed! Thank you  @michaelkrar  for the excellent work.

I've identified a few minor improvements that can be made and split them out into new issues. See https://github.com/orgs/vizhub-core/projects/2/views/1?filterQuery=assignee%3Amichaelkrar
